### ISSUE TITLE
Stopping Scylla while compaction is running may result in a crash

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#1900]
*Installation details*
Scylla version (or git commit hash): 1.4, 1.5-rc1

```
[54.193.122.180] [stdout] Nov 19 17:52:51 info   |  [shard 1] compaction_manager - Stopped
[54.193.122.180] [stdout] Nov 19 17:52:51 info   |  [shard 3] compaction_manager - Stopped
[54.193.122.180] [stdout] Nov 19 17:52:51 info   |  [shard 8] compaction_manager - Stopped
[54.193.122.180] [stdout] Nov 19 17:52:51 info   |  [shard 5] compaction_manager - Stopped
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | Segmentation fault on shard 0.
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | Backtrace:
[54.193.122.180] [stdout] Nov 19 17:52:51 info   |  [shard 2] compaction_manager - Stopped
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x00000000004e2e6a
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x00000000004e2f13
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x00007f5b91a2e0ff
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x000000000054c10c
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x000000000055e00e
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x000000000056d8ed
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x00000000004e4ace
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x000000000052f872
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x000000000059b26c
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x000000000047637c
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x00007f5b9167eb14
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0x00000000004dffa0
[54.193.122.180] [stdout] Nov 19 17:52:51 info   | 0xffffffffffffffff
```

```
print_backtrace_safe at /usr/src/debug/scylla-1.4.1/seastar/core/reactor.cc:2252
 (inlined by) print_with_backtrace at /usr/src/debug/scylla-1.4.1/seastar/core/reactor.cc:3087
sigsegv_action at /usr/src/debug/scylla-1.4.1/seastar/core/reactor.cc:3091
 (inlined by) operator() at /usr/src/debug/scylla-1.4.1/seastar/core/reactor.cc:3067
 (inlined by) _FUN at /usr/src/debug/scylla-1.4.1/seastar/core/reactor.cc:3063
?? ??:0
void promise<>::set_value<>() at /usr/src/debug/scylla-1.4.1/seastar/core/future.hh:472
basic_semaphore<semaphore_default_exception_factory>::signal(unsigned long) at /usr/src/debug/scylla-1.4.1/seastar/core/semaphore.hh:219
basic_semaphore<semaphore_default_exception_factory>::signal(unsigned long) at /usr/src/debug/scylla-1.4.1/seastar/core/fstream.cc:179
 (inlined by) file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}::operator()() const at /usr/src/debug/scylla-1.4.1/seastar/core/fstream.cc:179
 (inlined by) future<>::finally_body<file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}, false>::operator()(future<>&&) at /usr/src/debug/scylla-1.4.1/seastar/core/future.hh:977
 (inlined by) future<> futurize<future<> >::apply<future<>::finally_body<file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}, false>, future<> >(future<>::finally_body<file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}, false>&&, future<>&&) at /usr/src/debug/scylla-1.4.1/seastar/core/future.hh:1248
 (inlined by) _ZZN6futureIJEE12then_wrappedINS0_12finally_bodyIZZN19file_data_sink_impl3putE16temporary_bufferIcEENUlvE_clEvEUlvE_Lb0EEES0_EET0_OT_ENUlSB_E_clI12future_stateIJEEEEDaSB_ at /usr/src/debug/scylla-1.4.1/seastar/core/future.hh:887
 (inlined by) _ZN12continuationIZN6futureIJEE12then_wrappedINS1_12finally_bodyIZZN19file_data_sink_impl3putE16temporary_bufferIcEENUlvE_clEvEUlvE_Lb0EEES1_EET0_OT_EUlSC_E_JEE3runEv at /usr/src/debug/scylla-1.4.1/seastar/core/future.hh:390
reactor::run_tasks(circular_buffer<std::unique_ptr<task, std::default_delete<task> >, std::allocator<std::unique_ptr<task, std::default_delete<task> > > >&) at /usr/src/debug/scylla-1.4.1/seastar/core/reactor.cc:1924
reactor::run() at /usr/src/debug/scylla-1.4.1/seastar/core/reactor.cc:2356
app_template::run_deprecated(int, char**, std::function<void ()>&&) at /usr/src/debug/scylla-1.4.1/seastar/core/app-template.cc:131
main at /usr/src/debug/scylla-1.4.1/main.cc:663
```

Refs #1874, #1891 

Could be related to #990

**Comment by glommer** [comment]
@tgrabiec do we already have any idea about what could be causing this ?

**Comment by tgrabiec** [comment]
@glommer It very much looks like #990, which happens when you abandon a writer without closing it, typically when exception is thrown which interrupts the writer. I wasn't able to reproduce this particular scenario (crash while stopping compaction). It seems very likely that the writer is interrupted by exception, but the fact that we get an exception in the first when stopping indicates there is a bug in the shutdown sequence, maybe we're not waiting for all writers and they run into undefined behavior.

**Comment by roydahan** [comment]
I think the issue was just reproduced on longevity-1.6-1TB test.

This is the relevant section in the log:
`2017-02-02 17:09:42,433 nemesis          L0161 INFO | sdcm.nemesis.ChaosMonkey: StopStartService Node longevity-1.6-1TB-scylla-db-node-c1dc9d95-3 [54.84.58.48 | 172.30.0.176] (seed: False)
2017-02-02 17:09:42,433 remote           L0862 DEBUG| Remote [centos@54.84.58.48]: Running 'sudo systemctl stop scylla-server.service'
2017-02-02 17:09:42,434 remote           L0094 INFO | [54.84.58.48] Running '/bin/ssh  -a -x  -o ControlPath=/var/tmp/ssh-masterpD5j1Z/socket -o StrictHostKeyChecking=no -o UserKnownHostsFile=/var/tmp/tmpXpWLZ_ -o BatchMode=yes -o ConnectTimeout=300 -o ServerAliveInterval=300 -l centos -p 22 -i /var/tmp/longevity-1.6-1TB-sct-44ce0304.pem 54.84.58.48 "sudo systemctl stop scylla-server.service"'
2017-02-02 17:09:42,606 remote           L0194 DEBUG| [52.207.52.165] [stdout] total,     624110384,   28927,   28927,   28927,     3.4,     0.9,    13.1,    19.8,    29.0,    39.9,18086.6,  0.00414,      0,      0,       0,       0,       0,       0
2017-02-02 17:09:42,630 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | Stopping Scylla JMX...
2017-02-02 17:09:42,980 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 notice | scylla-jmx.service: main process exited, code=exited, status=143/n/a
2017-02-02 17:09:42,980 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | Stopped Scylla JMX.
2017-02-02 17:09:42,980 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 notice | Unit scylla-jmx.service entered failed state.
2017-02-02 17:09:42,981 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 warning| scylla-jmx.service failed.
2017-02-02 17:09:42,981 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | Stopping Scylla Server...
2017-02-02 17:09:42,981 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 0] compaction_manager - Asked to stop
2017-02-02 17:09:42,981 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 0] compaction_manager - Stopped
2017-02-02 17:09:42,982 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 3] compaction_manager - Asked to stop
2017-02-02 17:09:43,056 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 3] compaction_manager - Stopped
2017-02-02 17:09:43,057 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 6] compaction_manager - Asked to stop
2017-02-02 17:09:43,057 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 11] compaction_manager - Asked to stop
2017-02-02 17:09:43,057 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 12] compaction_manager - Asked to stop
2017-02-02 17:09:43,057 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 11] compaction_manager - Stopped
2017-02-02 17:09:43,057 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 4] compaction_manager - Asked to stop
2017-02-02 17:09:43,058 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 13] compaction_manager - Asked to stop
2017-02-02 17:09:43,058 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 4] compaction_manager - Stopped
2017-02-02 17:09:43,058 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 10] compaction_manager - Asked to stop
2017-02-02 17:09:43,058 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 9] compaction_manager - Asked to stop
2017-02-02 17:09:43,058 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 9] compaction_manager - Stopped
2017-02-02 17:09:43,059 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 6] compaction_manager - Stopped
2017-02-02 17:09:43,059 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 7] compaction_manager - Asked to stop
2017-02-02 17:09:43,059 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 13] compaction_manager - Stopped
2017-02-02 17:09:43,059 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 1] compaction_manager - Asked to stop
2017-02-02 17:09:43,059 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 8] compaction_manager - Asked to stop
2017-02-02 17:09:43,060 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 12] compaction_manager - Stopped
2017-02-02 17:09:43,060 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 2] compaction_manager - Asked to stop
2017-02-02 17:09:43,060 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 5] compaction_manager - Asked to stop
2017-02-02 17:09:43,060 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 10] compaction_manager - Stopped
2017-02-02 17:09:43,060 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 7] compaction_manager - compaction info: Compaction for keyspace1/standard1 was stopped due to shutdown.
2017-02-02 17:09:43,061 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 7] compaction_manager - Stopped
2017-02-02 17:09:43,061 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 2] compaction_manager - compaction info: Compaction for keyspace1/standard1 was stopped due to shutdown.
2017-02-02 17:09:43,061 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 2] compaction_manager - Stopped
2017-02-02 17:09:43,061 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 8] compaction_manager - compaction info: Compaction for keyspace1/standard1 was stopped due to shutdown.
2017-02-02 17:09:43,062 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   |  [shard 8] compaction_manager - Stopped
2017-02-02 17:09:43,062 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | Segmentation fault on shard 1.
2017-02-02 17:09:43,062 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | Backtrace:
2017-02-02 17:09:43,062 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | 0x00000000004e047a
2017-02-02 17:09:43,062 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | 0x00000000004e0523
2017-02-02 17:09:43,062 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | 0x00007f6e8cda70ff
2017-02-02 17:09:43,063 remote           L0194 DEBUG| [54.84.58.48] [stdout] Feb 02 17:09:42 info   | 0x00006010003a115f
2017-02-02 17:09:43,101 remote           L0194 DEBUG| [54.175.105.179] [stdout] Feb 02 17:09:43 info   |  [shard 13] compaction - Compacted 4 sstables to [/var/lib/scylla/data/keyspace1/standard1-42d76550e94011e69c39000000000002/keyspace1-standard1-ka-3989-Data.db:level=0, ]. 365986614 bytes to 365986614 (~100% of original) in 12096ms = 28.8552MB/s. ~351616 total partitions merged to 350898.
2017-02-02 17:09:44,619 remote           L0194 DEBUG| [54.175.105.179] [stdout] Feb 02 17:09:44 info   |  [shard 13] compaction - Compacted 4 sstables to [/var/lib/scylla/data/keyspace1/standard1-42d76550e94011e69c39000000000002/keyspace1-standard1-ka-3933-Data.db:level=0, ]. 2023660933 bytes to 2023660933 (~100% of original) in 43946ms = 43.9156MB/s. ~1940992 total partitions merged to 1940231.
2017-02-02 17:09:45,095 remote           L0194 DEBUG| [54.175.105.179] [stdout] Feb 02 17:09:45 info   |  [shard 8] compaction - Compacted 4 sstables to [/var/lib/scylla/data/keyspace1/standard1-42d76550e94011e69c39000000000002/keyspace1-standard1-ka-3984-Data.db:level=0, ]. 1997404451 bytes to 1997404451 (~100% of original) in 39606ms = 48.0956MB/s. ~1915776 total partitions merged to 1915057.
2017-02-02 17:09:51,652 remote           L0194 DEBUG| [54.175.105.179] [stdout] Feb 02 17:09:51 info   |  [shard 5] compaction - Compacted 4 sstables to [/var/lib/scylla/data/keyspace1/standard1-42d76550e94011e69c39000000000002/keyspace1-standard1-ka-4037-Data.db:level=0, ]. 1690663366 bytes to 1690663366 (~100% of original) in 27146ms = 59.3952MB/s. ~1621760 total partitions merged to 1620962.`

There is a coredump on the relevant db-node.

**Comment by tgrabiec** [comment]
The coredump file seems broken:

```
BFD: Warning: /var/lib/systemd/coredump/core.scylla.994.e014d29f6d81419d876ef23a5ff5357d.30534.1486055382000000 is truncated: expected core file size >= 115680100352, found: 88074506240.
```

Perhaps because `scylla-server.service` killed it when stopping timed out:

```
Feb 02 17:09:42 ip-172-30-0-176 kernel: scylla[30536]: segfault at 6010003a1160 ip 00006010003a1160 sp 00007f6e883f8058 error 15
...
Feb 02 17:11:13 ip-172-30-0-176 systemd[1]: scylla-server.service stop-sigterm timed out. Killing.
Feb 02 17:11:14 ip-172-30-0-176 systemd[1]: scylla-server.service: main process exited, code=killed, status=11/SEGV
Feb 02 17:11:14 ip-172-30-0-176 systemd-coredump[40890]: Process 30534 (scylla) of user 994 dumped core.
                                                         
                                                         Stack trace of thread 30536:
                                                         #0  0x00006010003a1160 n/a (n/a)
Feb 02 17:11:14 ip-172-30-0-176 systemd[1]: Stopped Scylla Server.
Feb 02 17:11:14 ip-172-30-0-176 systemd[1]: Unit scylla-server.service entered failed state.
Feb 02 17:11:14 ip-172-30-0-176 systemd[1]: scylla-server.service failed.

```

```
(gdb) p local_engine
Cannot access memory at address 0x7f6e908191e8
Cannot access memory at address 0x7f6e908191e0
Cannot find thread-local storage for LWP 30536, executable file /usr/lib/debug/usr/bin/scylla.debug:
Cannot find thread-local variables on this target

```

**Comment by raphaelsc** [comment]
This can be closed after 0977f4fdf8f1334d4544015c65a14f3977979f91
